rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+## <a id="p5" /> (P5) Unify static and dynamic versions of StableHLO operations
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** There are several operations in the StableHLO dialect which
+provide dynamic versions of existing StableHLO operations. For example, `PadOp`
+defines `edge_padding_low`, `edge_padding_high` and `interior_padding` as static
+attributes, whereas `DynamicPadOp` has the same contract except that those
+arguments are dynamic values.
+
+```mlir
+// Static version - in PadOp, padding options are represented as attributes.
+// %arg0 is the operand, %arg1 is the padding value.
+%0 = stablehlo.pad %arg0, %arg1, low = [1, 1], high = [1, 1], interior = [0, 0]
+  : (tensor<16x16xf32>, tensor<f32>) -> tensor<20x20xf32>
+
+// Dynamic version - in DynamicPadOp, padding options are represented as values.
+// The rest of the contract is the same.
+%2 = stablehlo.dynamic_pad %arg0, %arg1, %arg2, %arg3, %arg4 :
+  : (tensor<?x?xf32>, tensor<f32>, tensor<2xindex>, tensor<2xindex>, tensor<2xindex>) -> tensor<?x?xf32>
+```
+
+**Proposal:** This RFC proposes to take the 16 operations from [P4](#p4)
+(7 operations with "load-bearing" result types and 9 operations with
+size-related attributes) and generalize them to express both static and dynamic
+semantics.
+
+More specifically, the specification of these operations would change as follows
+(with special cases for `broadcast_in_dim` and `dynamic_slice` which are
+discussed separately below):
+
+1) For operations with "load-bearing" output types, add a new `output_shape`
+   input that has type "variadic number of 0-dimensional tensors of integer
+   type" and add a constraint `output_shape = shape(result)`. If the operation
+   has a variadic number of results, that would be `output_shapes`,
+   "variadic number of variadic number of ..." and
+   `output_shapes = shape(results...)`.
+2) For operations with size-related inputs, generalize them from
+   "1-dimensional tensor constant of type si64" to "variadic number of
+   0-dimensional tensors of integer type". The notation for existing
+   constraints pertaining to generalized inputs doesn't need to change
+   per [P3](#p3).
+
+The implementation of the StableHLO dialect would change as well, building
+on the experience from MLIR upstream where operations like
+`tensor::ExtractSliceOp`, `tensor::InsertSliceOp`, `tensor::PadOp` and others
+support both static and dynamic arguments
+([example](https://github.com/llvm/llvm-project/blob/f8cf2105760d2dfd58de4ca1c459c5bb1e2802dc/mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td#L1161)
+). Note that this section is still using tensors to represent dynamic sizes,
+whereas MLIR upstream is using scalars. As mentioned in [F4](#f4), there's been
+community feedback that this design can be improved, and the next section
+proposes how an improved design could look like.
+
+```tablegen
+// Before: static version.
+def StableHLO_PadOp ... {
+  let arguments = (ins
+    HLO_Tensor:$operand,
```

Comment:
nit: I'm curious why these are named HLO_Tensor types and not StableHLO_Tensor?

Reply:
Because they are shared with multiple *HLO dialects, including CHLO and MHLO.

Reply:
Ahh ack, then super nit: Maybe rename to AnyHLO? Otherwise I would presume this links to basic HLO or whatever we call regular HLO now :)